File: android/SceneWorld001/app/src/main/java/z3d/display/role/Display3dMovie.java
package z3d.display.role;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import z3d.base.GroupBackFun;
import z3d.base.GroupItem;
import z3d.base.MeshData;
import z3d.base.Scene_data;
import z3d.base.SkinMeshBackFun;
import z3d.core.Context3D;
import z3d.display.Display3DSprite;
import z3d.display.interfaces.IBind;
import z3d.display.particle.CombineParticle;
import z3d.filemodel.GroupDataManager;
import z3d.filemodel.MeshDataManager;
import z3d.filemodel.ParticleManager;
import z3d.program.Shader3D;
import z3d.res.BaseRes;
import z3d.res.GroupRes;
import z3d.scene.Scene3D;
import z3d.vo.AnimData;
import z3d.vo.DualQuatFloat32Array;
import z3d.vo.Matrix3D;
import z3d.vo.SkinMesh;
import z3d.vo.Vector3D;

public class Display3dMovie extends Display3DSprite implements IBind {
    public static  String TAG="Display3dMovie";
    public SkinMesh skinMesh;
    public float fileScale;
    public HashMap animDic;
    public boolean meshVisible;
    public  String curentAction;
    public  String defaultAction="stand";
    public  int curentFrame;
    public  float actionTime;
    public int  completeState;
    protected HashMap _partDic;
    protected HashMap _partUrl ;
    public  Display3dMovie(Scene3D val)
    {
        super(val);
        this.meshVisible=true;
        this.curentFrame=0;
        this.actionTime=0;
        this.completeState=0;
        this._partDic=new HashMap();
        this._partUrl=new HashMap();

    }
    public void  setRoleUrl(String url)
    {
        MeshDataManager.getInstance().getMeshData(url, new SkinMeshBackFun() {
            @Override
            public void Bfun(SkinMesh value) {
                skinMesh=value;
                fileScale=skinMesh.fileScale;
                animDic = skinMesh.animDic;
                updateMatrix();

            }
        }, 1);

    }
    @Override
    public void upData() {
        if(this.skinMesh==null||this.scene3d==null){
            return;
        }
        //this.updateBind();
        if(this.meshVisible){
            for (int i = 0; i < this.skinMesh.meshAry.size(); i++) {
                this.updateMaterialMesh(this.skinMesh.meshAry.get(i));
            }
        }

    }
    public void   updateFrame(float t){
        this.actionTime+=t;
        if(this.skinMesh==null){
            return;
        }
        AnimData  animData=_getCurentAnimData();
        if (animData==null) {
            return;
        }
        this.curentFrame=(int)(this.actionTime/(Scene_data.frameTime*1.5) );
        if (this.curentFrame >= animData.matrixAry.size()) {
            if (this.completeState == 0) {
                this.actionTime = 0;
                this.curentFrame = 0;
            } else if (this.completeState == 1) {
                this.curentFrame =(int) animData.matrixAry.size() - 1;
            } else if (this.completeState == 2) {
                this.curentFrame = 0;
                this.completeState = 0;
            } else if (this.completeState == 3) {
            }
        }

    }
    protected void  updateMaterialMesh(MeshData mesh)
    {
        if (mesh.material==null ) {
            Log.d(TAG, "没有: ");
            return;
        }
        if (mesh.material.shader==null ) {
            Log.d(TAG, "没有: ");
            return;
        }
        this.shader3D=mesh.material.shader;
        Context3D ctx=this.scene3d.context3D;
        ctx.setDepthTest(true);
        ctx.setWriteDepth(true);
        ctx.setProgame(this.shader3D.program);
        this.setVc();
        this.setMaterialTexture(mesh.material,mesh.materialParam);
        this.setMeshVc(mesh);
        ctx.setVa(this.shader3D,"pos",3,mesh.vertexBuffer);
        ctx.setVa(this.shader3D,"v2Uv",2,mesh.uvBuffer);
        ctx.setVa(this.shader3D,"boneID",4,mesh.boneIdBuffer);
        ctx.setVa(this.shader3D,"boneWeight",4,mesh.boneWeightBuffer);
        ctx.drawCall(mesh.indexBuffer,mesh.treNum);

//        Log.d(TAG,"-----");

    }

    protected void setVc()
    {
        Context3D ctx=this.scene3d.context3D;
        this.modeMatrix.identity();
        this.modeMatrix.appendScale(0.1f,0.1f,0.1f);
        ctx.setVcMatrix4fv(this.shader3D, Shader3D.vpMatrix3D,this.scene3d.camera3D.modelMatrix.m);
        ctx.setVcMatrix4fv(this.shader3D,Shader3D.posMatrix,this.modeMatrix.m);

    }
    private AnimData  _getCurentAnimData(){
        this.curentAction="walk";
        this.curentAction=defaultAction;
        AnimData  animData = null;
        if (this.animDic.containsKey(this.curentAction)) {
            animData = (AnimData)this.animDic.get(this.curentAction);
        } else if (this.animDic.containsKey(this.defaultAction)) {
            animData =(AnimData) this.animDic.get(this.defaultAction);
        }
        return animData;
    }

    private  void setMeshVc(MeshData mesh)
    {
        AnimData  animData=_getCurentAnimData();
//        this.curentFrame=0;
        DualQuatFloat32Array dualQuatFrame =  animData.boneQPAry.get(mesh.uid).get(this.curentFrame);
        Context3D ctx=this.scene3d.context3D;
        ctx.setVc4fv(this.shader3D,"boneQ",54, dualQuatFrame.boneQarrrBuff);
        ctx.setVc3fv(this.shader3D,"boneD",54, dualQuatFrame.boneDarrBuff);

    }

    public void addPart(String key,String bindSocket, String url){
        if ( this._partUrl.get(key)==url) {//如果相同则返回
            return;
        } else if (this._partUrl.containsKey(key)) {//如果不同则先移除
            this.removePart(key);
        }
        if (!this._partDic.containsKey(key)) {
            this._partDic.put(key,new ArrayList<>());
        }
        this._partUrl.put(key,url);
        ArrayList ary= (ArrayList)this._partDic.get(key);
        GroupDataManager.getInstance().getGroupData( url, new GroupBackFun() {
            @Override
            public void Bfun(GroupRes groupRes) {
                loadPartRes(bindSocket, groupRes, ary);
            }
        });
    }
    private void  loadPartRes(String bindSocket,GroupRes groupRes,ArrayList ary){

        for (int i = 0; i < groupRes.dataAry.size(); i++) {
            GroupItem item   = groupRes.dataAry.get(i);

            Vector3D posV3d = null;
            Vector3D rotationV3d = null;
            Vector3D scaleV3d = null;
            if (item.isGroup) {
                posV3d = new Vector3D(item.x, item.y, item.z);
                rotationV3d = new Vector3D(item.rotationX, item.rotationY, item.rotationZ);
                scaleV3d = new Vector3D(item.scaleX, item.scaleY, item.scaleZ);
            }

            if (item.types == BaseRes.SCENE_PARTICLE_TYPE) {
                CombineParticle particle = ParticleManager.getInstance().getParticleByte(Scene_data.fileRoot + item.particleUrl);
                ary.add(particle);
                particle.setBindTarget(this);
                particle.bindSocket = bindSocket;
                particle.dynamic = true;
                ParticleManager.getInstance().addParticle(particle);
                if (item.isGroup) {
                    particle.setGroup(posV3d, rotationV3d, scaleV3d);
                }
            } else if (item.types == BaseRes.PREFAB_TYPE) {
                Display3DSprite display   = new Display3DSprite(scene3d);
                display.setObjUrl(item.objUrl);
                display.setMaterialUrl(item.materialUrl, item.materialInfoArr);
                display.dynamic = true;
                ary.add(display);
                display.setBind(this, bindSocket);
                scene3d.addSpriteDisplay(display);
                if (item.isGroup) {
                    display.setGroup(posV3d, rotationV3d, scaleV3d);
                }

            }

        }

    }
    private void removePart(String key) {
    }


    @Override
    public void getSocket(String socketName, Matrix3D resultMatrix) {

    }

    @Override
    public int getSunType() {
        return 0;
    }
}

File: metalios/iosgl/iosgl/pan3d/display/role/Display3dMovie.m
//
//  Display3dMovie.m
//  iosgl
//
//  Created by zhao on 2/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Display3dMovie.h"
#import "SkinMesh.h"
#import "Scene_data.h"
#import "AnimData.h"
#import "Context3D.h"
#import "Scene3D.h"
#import "GroupItem.h"
#import "TextureRes.h"
#import "BoneSocketData.h"
#import "DualQuatFloat32Array.h"
#import "ProgrmaManager.h"
#import "MaterialAnimShader.h"
#import "MeshDataManager.h"
#import "Display3DSprite.h"
#import "TextureManager.h"
#import "GroupDataManager.h"
#import "MtlMoveDisplayType.h"
#import "DynamicTexItem.h"

@interface Display3dMovie()<IBind>
@property(nonatomic,strong)NSString*  meshUrl;

@property(nonatomic,strong)NSMutableDictionary*  partDic;
@property(nonatomic,strong)NSMutableDictionary*  partUrl;
@property(nonatomic,strong)NSMutableDictionary*  preLoadActionDic;
@property(nonatomic,strong)NSMutableDictionary*  waitLoadActionDic;
 

@property(nonatomic,assign)int  completeState ;
@property(nonatomic,assign)int  curentFrame;
@property(nonatomic,assign)float actionTime;
@property(nonatomic,assign)float fileScale;
@property(nonatomic,assign)BOOL meshVisible;

@property(nonatomic,strong)Skill* skillVo;

@end
@implementation Display3dMovie
- (instancetype)init:value
{
    /*
     [0]    (null)    @"attack_03" : (no summary)
     [1]    (null)    @"stand" : (no summary)
     [2]    (null)    @"walk" : (no summary)
     [3]    (null)    @"stand_mount_01" : (no summary)
     [4]    (null)    @"attack_01" : (no summary)
     [5]    (null)    @"jump" : (no summary)
     [6]    (null)    @"death" : (no summary)
     [7]    (null)    @"injured" : (no summary)
     [8]    (null)    @"attack_010" : (no summary)
     [9]    (null)    @"attack_02" : (no summary)
     [10]    (null)    @"m_attack_01" : (no summary)
     [11]    (null)    @"m_attack_02" : (no summary)
     [12]    (null)    @"walk_mount_01" : (no summary)
     [13]    (null)    @"m_attack_03" : (no summary)
     [14]    (null)    @"attack_020" : (no summary)
     [15]    (null)    @"m_attack_04" : (no summary)
     */
    self = [super init:value];
    if (self) {
        Display3dMovie* this=self;
        
        this.meshVisible=YES;
        this.defaultAction= @"stand";
        this.partDic = [[NSMutableDictionary alloc]init];
        this.partUrl =[[NSMutableDictionary alloc]init];
        this.preLoadActionDic = [[NSMutableDictionary alloc]init];
        this.waitLoadActionDic =[[NSMutableDictionary alloc]init];
        this.actionTime=0;
        
    
   
    }
    return self;
}
 

-(void)onMeshLoaded;
{
    
}
- (void)setVa;
{
    
}
-(void)clearMesh;
{
    
}
- (int)getSunType
{
    return 1;
}
- (void)getSocket:(NSString *)socketName resultMatrix:(Matrix3D *)resultMatrix
{
    Display3dMovie* this =self;
    [resultMatrix identity];
    if (!this.skinMesh) {
        [resultMatrix append:this.posMatrix3d];
        return;
    } else if (!this.skinMesh.boneSocketDic[socketName]) {
        if ( [socketName isEqualToString:@"none"]) {
            [resultMatrix appendTranslation:this.x y:this.y z:this.z];
        } else {
            [resultMatrix append:this.posMatrix3d];
        }
        return;
    }
    BoneSocketData* boneSocketData   = this.skinMesh.boneSocketDic[socketName];
    Matrix3D* testmatix  ;
    int index    = boneSocketData.index;
    testmatix = [this getFrameMatrix:index];
    [resultMatrix appendScale:1/this.scaleX y:1/this.scaleY z:1/self.scaleZ];
    [resultMatrix appendRotation:boneSocketData.rotationX axis:Vector3D.X_AXIS];
    [resultMatrix appendRotation:boneSocketData.rotationY axis:Vector3D.Y_AXIS];
    [resultMatrix appendRotation:boneSocketData.rotationZ axis:Vector3D.Z_AXIS];
    [resultMatrix appendTranslation:boneSocketData.x y:boneSocketData.y z:boneSocketData.z];
    if(testmatix){
        [resultMatrix append:this.skinMesh.bindPosInvertMatrixAry[index]];
        [resultMatrix append:testmatix];
    }
    [resultMatrix append:this.posMatrix3d];
 
  
}
-(Matrix3D*)getFrameMatrix:(int)index;
{
     Display3dMovie* this =self;
    if (this.animDic[this.curentAction]) {
        AnimData* animData   = this.animDic[this.curentAction];
        return animData.matrixAry[this.curentFrame][index];
    } else if (this.animDic[this.defaultAction]) {
        AnimData* animData  = this.animDic[this.defaultAction];
        return animData.matrixAry[this.curentFrame][index];
    }

    return nil;
}

-(void)setRoleUrl:(NSString*)value;
{
     Display3dMovie* this =self;
    
    [self.scene3D.meshDataManager getMeshData:value fun:^(SkinMesh * _Nonnull skinMesh) {
        this.skinMesh=skinMesh;
        this.fileScale=skinMesh.fileScale;
        this.animDic = skinMesh.animDic;
        [this onMeshLoaded];
        for (int i = 0; i < this.skinMesh.meshAry.count; i++) {
            [skinMesh.meshAry[i] upToGpu];
        }
        [this updateMatrix];
    } batchNum:1];
}


- (void)upFrame;
{
    Display3dMovie* this=self;
    if(!this.skinMesh){
        return;
    }
    [this updateBind];
    if(this.meshVisible){
        for (int i = 0; i < self.skinMesh.meshAry.count; i++) {
            [this updateMaterialMesh:this.skinMesh.meshAry[i]];
        }
    }
    
}
-(BOOL)play:(NSString*)action
{
    return [self play:action completeState:0 needFollow:true];
}
/*
 播放动作  播放完状态  0c持续  1播完听 2 播放返回
 */
-(BOOL)play:(NSString*)action completeState:(int)completeState needFollow:(BOOL)needFollow;
{
    Display3dMovie* this=self;
    if (this.curentAction == action) {
        return YES;
    }
    this.curentAction = action;
    this.completeState = completeState;
    this.actionTime = 0;
    [this updateFrame:0];
    if ([this.animDic valueForKey:action]) {
        return YES;
    } else {
        return NO;
    }
}
 
 
-(void)setMeshVc:(MeshData*)mesh redEncoder:(id<MTLRenderCommandEncoder>)renderEncoder
{
    Display3dMovie* this=self;
 
    AnimData* animData;
    if (this.animDic[this.curentAction]) {
        animData = this.animDic[this.curentAction];
    } else if (this.animDic[this.defaultAction]) {
        animData = this.animDic[this.defaultAction];
    } else {
        return;
    }

    
    DualQuatFloat32Array* dualQuatFrame = animData.boneQPAry[mesh.uid][self.curentFrame];
    [renderEncoder setVertexBuffer: dualQuatFrame.mtkquatArr   offset:0   atIndex:5];
    [renderEncoder setVertexBuffer: dualQuatFrame.mtkposArr   offset:0   atIndex:6];
}

 
 
-(void)updateMatrix;
{
    [super updateMatrix];
    [self.posMatrix3d prependScale:self.fileScale y:self.fileScale z:self.fileScale];
 
}
- (void)setVcMtl:(id<MTLRenderCommandEncoder>)renderEncoder {
    Matrix3D* posMatrix =[[Matrix3D alloc]init];
    LineMatrixRoleView matrix = {[self.scene3D.camera3D.modelMatrix getMatrixFloat4x4], [posMatrix getMatrixFloat4x4]};
    [renderEncoder setVertexBytes:&matrix
                           length:sizeof(matrix)
                          atIndex:0];
}

/*
 部分mesh对象渲染
 */
-(void)updateMaterialMesh:(MeshData*)mesh;
{
    Display3dMovie* this=self;
    if (!mesh.material) {
        return;
    }
    
    this.shader3d=mesh.material.shader;
    
    
    id<MTLRenderCommandEncoder> renderEncoder=this.scene3D.context3D.renderEncoder;
    
    [self.shader3d mtlSetProgramShader];
    
    [self setVcMtl:renderEncoder];
    
    
    [renderEncoder setCullMode:0];
    
    [self setMeshVc:mesh redEncoder:renderEncoder];
    
    [renderEncoder setVertexBuffer: mesh.mtkvertices  offset:0  atIndex:1];
    [renderEncoder setVertexBuffer: mesh.mtkuvs  offset:0  atIndex:2];
    [renderEncoder setVertexBuffer: mesh.mtkboneId  offset:0  atIndex:3];
    [renderEncoder setVertexBuffer: mesh.mtkboneWeight   offset:0   atIndex:4];
    
    
    [this setMaterialTexture:mesh.material mp:mesh.materialParam];
    
    
    [renderEncoder drawIndexedPrimitives:MTLPrimitiveTypeTriangle
                              indexCount: mesh.mtkindexCount
                               indexType:MTLIndexTypeUInt32
                             indexBuffer: mesh.mtkindexs
                       indexBufferOffset:0];
    
    
}


-(void)setMaterialTexture:(Material*)material  mp:(MaterialBaseParam*)mp;
{
 
    NSArray<TexItem*>* texVec  = mp.material.texList;
    TexItem* texItem;
    for (int i   = 0; i < texVec.count; i++) {
        texItem=texVec[i];
        if (texItem.isDynamic) {
            continue;
        }
        if (texItem.type == TexItem.LIGHTMAP) {
        }
        else if (texItem.type == TexItem.LTUMAP && [Scene_data default].pubLut ) {
            NSLog(@"TexItem.LTUMAP)");
        }
        else if (texItem.type == TexItem.CUBEMAP) {
            if (material.useDynamicIBL) {// && _reflectionTextureVo) {
                NSLog(@"TexItem.useDynamicIBL)");
            } else {
                if([Scene_data default].skyCubeTexture){
                  
                }
            }
        }
        else if (texItem.type == 0) {
         
            
        }
    }
    NSArray<DynamicTexItem*>* texDynamicVec  =( NSArray<DynamicTexItem*>*) mp.dynamicTexList;
    for (int i   = 0; i < texDynamicVec.count; i++) {
        texItem=texDynamicVec[i].target;
        if(texItem ){
            
            if(texItem.isMain){
                id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
                [renderEncoder setFragmentTexture:texDynamicVec[i].textureRes.mtlTexture
                                          atIndex:0];
                
              
            }
 
            
        }
    }
    
}
/*
 时间轴更新
 */
- (void)updateFrame:(float)t;
{
      Display3dMovie* this=self;
    if(!self.skinMesh){
        return;
    }
    this.actionTime+=t;
    NSString* actionKey;
    if(this.curentAction&&self.animDic[this.curentAction]){
        actionKey = this.curentAction;
    }else if ( this.animDic[this.defaultAction]){
        actionKey = this.defaultAction;
    }else{
        return;
    }
    AnimData* animData=this.animDic[actionKey];
    this.curentFrame=(int)(this.actionTime/([Scene_data default].frameTime*1.5) );
 
    if (this.curentFrame >= animData.matrixAry.count) {
        if (this.completeState == 0) {
            this.actionTime = 0.0f;
            this.curentFrame = 0;
        } else if (this.completeState == 1) {
            this.curentFrame =(int) animData.matrixAry.count - 1;
        } else if (this.completeState == 2) {
            this.curentFrame = 0;
            this.completeState = 0;
            [this changeAction:this.curentAction];
        } else if (this.completeState == 3) {
        }
    }
}
-(void)changeAction:(NSString*)action;
{
    self.curentAction = self.defaultAction;
}
/*
 部位，路径，类型 1为粒子 0为其他
 */
-(void)addPart:(NSString*)key bindSocket:(NSString*)bindSocket url:(NSString*)url;
{
    Display3dMovie* this=self;
    if([this.partUrl[key] isEqual:url]){
        return;
    }else if ( [this.partUrl valueForKey:key]){
        [this.partUrl removeObjectForKey:key];
    }
    if ( [this.partDic valueForKey:key]){
        this.partDic[key] = [[NSMutableArray alloc]init];
    }
    this.partUrl[key] = url;
    NSMutableArray* ary  = this.partDic[key];
    
    [self.scene3D.groupDataManager getGroupData:[[Scene_data default]getWorkUrlByFilePath:url] Block:^(GroupRes *groupRes) {
        [this loadPartRes:bindSocket groupRes:groupRes ary:ary];
    }];
}
/*
 部件数据加载返回
 */
-(void)loadPartRes:(NSString*)bindSocket groupRes:(GroupRes*)groupRes ary:(NSMutableArray*)ary;
{
    Display3dMovie* this=self;
    for (int i = 0; i < groupRes.dataAry.count; i++) {
          GroupItem* item  = groupRes.dataAry[i];

          Vector3D* posV3d ;
          Vector3D* rotationV3d  ;
          Vector3D* scaleV3d  ;
          if (item.isGroup) {
              posV3d = [[Vector3D alloc]x:item.x y:item.y z:item.z];
              rotationV3d =[[Vector3D alloc]x:item.rotationX y:item.rotationY z:item.rotationZ];
              scaleV3d = [[Vector3D alloc]x:item.scaleX y:item.scaleY z:item.scaleZ];
          }
 
          if (item.types == SCENE_PARTICLE_TYPE) {
               CombineParticle*  particle =   [self.scene3D.particleManager   getParticleByte: item.particleUrl];
              [ary addObject:particle];
              particle.bindTarget = this;
              particle.bindSocket = bindSocket;
              particle.dynamic = YES;
             [this.scene3D.particleManager addParticle:particle];
              if (item.isGroup) {
                //  particle.setGroup(posV3d, rotationV3d, scaleV3d);
              }
          } else if (item.types == PREFAB_TYPE) {
              Display3DSprite *display=[[Display3DSprite alloc]init:self.scene3D];
              [display setObjUrl:item.objUrl];
              [display setMaterialUrl:item.materialUrl paramData:item.materialInfoArr];
              [ary addObject:display];
              [display setBind:this bindSocket:bindSocket];
              [this.scene3D addDisplay:display];
              if(item.isGroup){
                  [display setGroup:posV3d rotaion:rotationV3d scale:scaleV3d];
              }
           

          }

      }
  
}
-(void)playSkill:(Skill*)skill;
{
    [self.scene3D.skillManager  playSkill:skill];

    self.skillVo=skill;
 
}
 
@end

 
